Some AST and SIL infrastructure additions and improvements

eeckstein opened the pull request:
* replace `CanonicalType.objectType` and `CanonicalType.addressType` with `silType`
* add `AST.Type.loweredType` and `CanonicalType.loweredType`
* let `SubstitutionMap.replacementTypes` return AST types rather than optional SIL types
* add some use-list APIs
* add `SingleValueInstruction.replace(with:)` and use it throughout the optimizer
* add `CanonicalType.canBeClass` and move the implementation of `SIL.Type.canBeClass` to the AST Type
* SimplifyBuiltin: support metatype and function types when optimizing `Builtin.sizeof`/`alignof`/`strideof`/`destroyArray`

For details see the commit messages

eeckstein committed:
SIL: replace `CanonicalType.objectType` and `CanonicalType.addressType` with `silType`

`SIL.Type` itself has `objectType` and `addressType` to convert between both variants.

eeckstein committed:
SIL: add `AST.Type.loweredType` and `CanonicalType.loweredType`

eeckstein committed:
SIL: Let `SubstitutionMap.replacementTypes` return AST types rather than optional SIL types.

This is what the C++ SubstitutionMap does. One has to use `Type.loweredType` to get from the AST type to the SIL type.

eeckstein committed:
SIL: add some use-list APIs

* `users`: maps from a sequence of operands to a sequence of instructions
* `users(ofType:)` : as `users`, but filters users of a certain instruction type
* `Value.users`: = `Value.uses.users`

eeckstein committed:
Optimizer: add `SingleValueInstruction.replace(with:)` and use it throughout the optimizer

It replaces all uses and then erases the instruction.

eeckstein commented:
@swift-ci smoke test

eeckstein commented:
@swift-ci test

eeckstein commented:
@swift-ci test

slavapestov commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/AllocVectorLowering.swift:
```diff
@@ -259,7 +260,8 @@ private func createStackAllocatedVector(
   let builder = Builder(before: allocVectorBuiltin, context)
-  let elementType = allocVectorBuiltin.substitutionMap.replacementTypes[0]!
+  let function = allocVectorBuiltin.parentFunction
+  let elementType = allocVectorBuiltin.substitutionMap.replacementTypes[0].loweredType(in: function)
```
Maybe for these cases, add a `var replacementType` computed property which asserts that the length is exactly one and then returns the sole element?

slavapestov commented on SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyBuiltin.swift:
```diff
@@ -129,7 +129,7 @@ private extension BuiltinInst {
 
   func optimizeCanBeClass(_ context: SimplifyContext) {
-    guard let ty = substitutionMap.replacementTypes[0] else {
+    guard let ty = substitutionMap.replacementTypes[0].canonical.silType else {
```
Should the canBeClass query be on AST types instead?

eeckstein replied:
yes, that makes sense

slavapestov commented on SwiftCompilerSources/Sources/Optimizer/InstructionSimplification/SimplifyBuiltin.swift:
```diff
@@ -33,7 +33,7 @@ extension BuiltinInst : OnoneSimplifyable {
       case .DestroyArray:
-        if let elementType = substitutionMap.replacementTypes[0],
+        if let elementType = substitutionMap.replacementTypes[0].canonical.silType,
            elementType.isTrivial(in: parentFunction)
```
You need to lower the type here because it might be a function type or metatype, no?

eeckstein replied:
Exactly.
I don't use `loweredType`, yet, because I need to add a flag for getting the "maximally abstracted" lowered type. This is what IRGen does for the size/alignment/stride builtins. This makes a difference for metatypes.
`Builtin.size(Bool.Type.self) == 8`. Whereas the "regular" type lowering is a thin metatype which would have a size of 0.

slavapestov reviewed (commented):
Looks great!

eeckstein committed:
SIL: add `CanonicalType.canBeClass`

And move the implementation of `SIL.Type.canBeClass` to the AST Type. The SIL Type just calls the AST Type implementation.
Also rename `SIL.Type.canonicalASTType` -> `SIL.Type.astType`.

eeckstein committed:
SimplifyBuiltin: support metatype and function types when optimizing Builtin.sizeof/alignof/strideof/destroyArray

Get the maximally abstracted lowered type instead of bailing if the ast type isn't equal to the lowered SIL type.

eeckstein commented:
@swift-ci test
